internal/broker/broker.go

```diff
@@ -2,15 +2,738 @@
+		return &Broker{}, errors.New("cache path must be provided")
+	}
+
+	clientID := cfg.ClientID
+	issuerURL := cfg.IssuerURL
+	if issuerURL == "" || clientID == "" {
+		return &Broker{}, errors.New("issuer and client ID must be provided")
+	}
+
+	// Generate a new private key for the broker.
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		panic(fmt.Sprintf("could not create an valid rsa key: %v", err))
+	}
+
+	// Create provider
+	provider, err := oidc.NewProvider(context.TODO(), issuerURL)
+	if err != nil {
+		return &Broker{}, err
+	}
+	oidcCfg := oidc.Config{
+		ClientID:                   clientID,
+		InsecureSkipSignatureCheck: opts.skipJWTSignatureCheck,
+	}
+	oauthCfg := oauth2.Config{
+		ClientID: clientID,
+		Endpoint: provider.Endpoint(),
+		Scopes:   append([]string{oidc.ScopeOpenID, "profile", "email"}, opts.providerInfo.AdditionalScopes()...),
+	}
+	authCfg := authConfig{
+		provider:    provider,
+		providerURL: issuerURL,
+		cachePath:   cfg.CachePath,
+		oidcCfg:     oidcCfg,
+		oauthCfg:    oauthCfg,
+	}
+
+	// Set offline expiration
+	var offlineExpiration time.Duration
+	if cfg.OfflineExpiration != "" {
+		intValue, err := strconv.Atoi(cfg.OfflineExpiration)
+		if err != nil {
+			return &Broker{}, fmt.Errorf("could not parse offline expiration: %v", err)
+		}
+		offlineExpiration = time.Duration(intValue*24) * time.Hour
+	}
+
+	return &Broker{
+		providerInfo:      opts.providerInfo,
+		auth:              authCfg,
+		offlineExpiration: offlineExpiration,
+		privateKey:        privateKey,
+
+		skipUsernameValidation: opts.skipUsernameValidation,
+
+		currentSessions:        make(map[string]sessionInfo),
+		currentSessionsMu:      sync.RWMutex{},
+		userLastSelectedMode:   make(map[string]string),
+		userLastSelectedModeMu: sync.Mutex{},
+		isAuthenticatedCalls:   make(map[string]isAuthenticatedCtx),
+		isAuthenticatedCallsMu: sync.Mutex{},
+	}, nil
+}
+
+// NewSession creates a new session for the user.
+func (b *Broker) NewSession(username, lang, mode string) (sessionID, encryptionKey string, err error) {
+	defer decorate.OnError(&err, "could not create new session for user %q", username)
+
+	sessionID = uuid.New().String()
+	session := sessionInfo{
+		username: username,
+		lang:     lang,
+		mode:     mode,
+
+		authInfo:        make(map[string]any),
+		attemptsPerMode: make(map[string]int),
+	}
+
+	pubASN1, err := x509.MarshalPKIXPublicKey(&b.privateKey.PublicKey)
+	if err != nil {
+		return "", "", err
+	}
+
+	_, url, _ := strings.Cut(b.auth.providerURL, "://")
+	url = strings.ReplaceAll(url, "/", "_")
+	url = strings.ReplaceAll(url, ":", "_")
+	session.cachePath = filepath.Join(b.auth.cachePath, url, username+".cache")
+
+	b.currentSessionsMu.Lock()
+	b.currentSessions[sessionID] = session
+	b.currentSessionsMu.Unlock()
+
+	return sessionID, base64.StdEncoding.EncodeToString(pubASN1), nil
+}
+
+// GetAuthenticationModes returns the authentication modes available for the user.
```

Comment:
Yeah, this is something that I don't like in the code either. The way I was doing this was: to try to separate some of these specific methods to the providers (as they can be quite specific about this). However, since we don't support more than one right now I decided to leave it as is and then touch this back later, when necessary (since we might still find some other refactoring worth pieces later down the line).

My suggestion is: to create a card in our backlog so that we remember that this is prone to refactor when adding support for more providers, wdyt?